specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,246 @@
+
+All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.
+
+## `libp2p` Protocol Names
+
+A "Protocol Name" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. A client's supported protocol paths are negotiated by the `libp2p` stack at connection time; as such they are not part of individual message bodies.
+
+## RPC-Over-`libp2p`
+
+To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:
+
+```
+(
+    id: uint64
+    method_id: uint16
+    body: Request
+)
+```
+
+and their corresponding responses are wrapped in a "response" structure:
+
+```
+(
+    id: uint64
+    result: Response
+)
+```
+
+If an error occurs, a variant of the response structure is returned:
+
+```
+(
+    id: uint64
+    error: (
+        code: uint16
+        data: bytes
+    )
+)
+```
+
+The details of the RPC-Over-`libp2p` protocol are similar to [JSON-RPC 2.0](https://www.jsonrpc.org/specification). Specifically:
+
+1. The `id` member is REQUIRED.
+2. The `id` member in the response MUST be the same as the value of the `id` in the request.
+3. The `method_id` member is REQUIRED.
+4. The `result` member is required on success, and MUST NOT exist if there was an error.
+5. The `error` member is REQUIRED on errors, and MUST NOT exist if there wasn't an error.
+
+Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching.
+
+The "method ID" fields in the below messages refer to the `method` field in the request structure above.
+
+The first 1,000 values in `error.code` are reserved for system use. The following error codes are predefined:
+
+1. `0`: Parse error.
+2. `10`: Invalid request.
+3. `20`: Method not found.
+4. `30`: Server error.
+
+## Messages
+
+### Hello
+
+**Method ID:** `0`
+
+**Body**:
+
+```
+(
+    network_id: uint8
+    latest_finalized_root: bytes32
+    latest_finalized_epoch: uint64
+    best_root: bytes32
+    best_slot: uint64
+)
+```
+
+Clients exchange `hello` messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the `hello` message. In response, the receiving client MUST respond with its own `hello` message.
+
+Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:
+
+1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.
```

Comment:
You should consider spelling out network ID and chain ID as separate fields. Chain ID should be set to a fixed number "1" for ETH, and if others want to run their own chain they can change that ID.

Reply:
NetworkId vs ChainId +1.
Also, message body compression algorithm indicator.
Also, upgrade paths for SSZ (I get the feeling this might change on the wire)..maybe a sorted list of serialization method preferences, the highest mutual being selected?